uswds/uswds
# USWDS - Packages: usa-icon

## Related issue
Closes https://github.com/uswds/uswds/issues/4867

## Preview Link
Preview link: [usa-icon in Storybook](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-usa-icon-create-module/?path=/story/design-tokens-icons--icons)

## Problem statement
Users should be able to install the isolated `usa-icon` package and have access to all of the icon files they need to design and build with USWDS. Currently, the commands `gulp buildSprite`, `gulp buildSpriteStandalone`, and `gulp buildUSWDS` output to `./dist` and not `./packages/usa-icon/dist`, which means users need the full uswds package to access our icon library. 

## Solution
**Change destination of gulp tasks**
Updating the destination in our existing gulp tasks to `./packages/usa-icon/dist` will create and consolidate the correct icon assets inside the usa-icon package. Once created inside usa-icon, we can copy these consolidated assets into `./dist/img` (for backwards compatibility).

**Redefine and rename `gulp buildSpriteStandalone`**
Renaming `gulp buildSpriteStandalone` to `gulp buildIcons` and expanding its build to generate all icon assets in usa-icon gives us a simple way to assemble the usa-icon dist without needing to build all of USWDS.

## Tasks
- [x] The commands `gulp buildSprite` and `gulp buildUSWDS` should output the same assets to both `./dist` and `./packages/usa-icon/dist`.
    - Exception: in `gulp buildUSWDS`, non-icon assets like `hero.png` and `/favicons`were intentionally excluded from `./packages/usa-icon/dist` but will appear in `dist/img`.
- [x] The commands `gulp buildSprite` and `gulp buildUSWDS` should output the same assets that are currently built in the ./dist directory in the `develop` branch.
- [x] `gulp buildIcons` should output all icon .svg assets to both package dist and root dist
- [x] Unit tests check that the correct non-empty directories and files exist in usa-icon package dist

## Testing and review
To test:
1. Check file structure
    1. Open the `al-usa-icon-create-module` branch
    1. Run `rm -rf dist` to remove all previously built dist files.
    1. Run the commands below one at a time, running`rm -rf dist packages/usa-icon/dist` in between each to clear the built dist directories:
        - `gulp buildSprite`
        - `gulp buildIcons`
        - `gulp buildUSWDS`
     1. After each gulp command, confirm that the files inside `./packages/usa-icon/dist/img` match what is found in `./dist/img`
        - Exception: in `gulp buildUSWDS`, `hero.png` and `/favicons` were intentionally excluded from `./packages/usa-icon/dist` but will appear in `dist/img`.
     1. Additionally, after each gulp command, confirm that the files inside `./dist/img` in this branch match what is generated in `./dist/img` in the `develop` branch. (Reference screenshots included below for easy reference on file structure).
         - Exception: ~`gulp buildSpriteStandalone`~ `gulp buildIcons` will now output all icon .svg assets (See screenshots below)
1. Confirm icon appearance in Storybook
    1. Open [Storybook ](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-usa-icon-create-module/?path=/story/design-tokens-icons--icons) and confirm that icons appear as expected.
1. Run npm test
     1. Run `npm run test` from both the root directory and the `./packages/usa-icon` directory and confirm that there are no errors. 
     
## Reference
### Output to `./dist`
- `gulp buildSprite` (no change)
    
    **Before:**
    ![image](https://user-images.githubusercontent.com/93996430/180801195-c5267a00-ab9d-4385-86d9-044a0a660567.png)

     **After:**
    ![image](https://user-images.githubusercontent.com/93996430/181123710-695e1ba3-f226-4b20-afc9-5c303b889253.png)

- `gulp buildSpriteStandalone`/`gulp buildIcons` (Now. more icon assets are generated. The results are the same as `gulpbuildUSWDS` but without favicons or hero.png)
    
    **Before (`gulp buildSpriteStandalone`):**
    ![image](https://user-images.githubusercontent.com/93996430/180800705-063634f8-db42-4487-b0fa-4f44c1d435b0.png)
    
     **After (`gulp buildIcons`):**
    ![image](https://user-images.githubusercontent.com/93996430/181624548-7a180447-87b3-4ba4-b323-8fba18d1b351.png)

- `gulp buildUSWDS` (No change)

    **Before:**
    ![image](https://user-images.githubusercontent.com/93996430/180801687-7f0cd835-cfbc-44ed-81ac-37ea0e26870b.png)
    
     **After:**
    ![image](https://user-images.githubusercontent.com/93996430/181625074-53f85932-d388-4acf-bcc5-01012749fc0b.png)

---

Before opening this PR, make sure you’ve done whichever of these applies to you:
